limits/filetransfer/limits.go

```diff
@@ -0,0 +1,392 @@
+		uint64(deviceCounters.bytesTransferredAvg1m) >= p.limits.FileTransfer.Counters.MaxBytesTxPerMinute {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesTransferred < math.MaxUint64-n {
+			p.counters.bytesTransferred += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BytesReceived(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCounters.bytesReceived < math.MaxUint64-n {
+			deviceCounters.bytesReceived += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesRxPerMinute > 0 &&
+		uint64(deviceCounters.bytesReceivedAvg1m) >= p.limits.FileTransfer.Counters.MaxBytesRxPerMinute {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesReceived < math.MaxUint64-n {
+			p.counters.bytesReceived += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BelowMaxAllowedFileSize(offset int64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	if offset < 0 {
+		return true
+	}
+	if p.limits.FileTransfer.MaxFileSize > 0 &&
+		uint64(offset) >= p.limits.FileTransfer.MaxFileSize {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & syscall.S_ISUID) != 0 {
+		mode &= os.ModePerm
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode |= os.ModeSetuid
+	} else {
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode &= os.ModePerm
+	}
+
+	if p.limits.FileTransfer.PreserveMode {
+		return os.Chmod(path, mode)
+	} else {
+		return nil
+	}
+}
+
+func (p *Permit) PreserveOwnerGroup(path string, uid int, gid int) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	forcedSet := false
+	if p.limits.FileTransfer.OwnerPut != "" {
+		u, err := user.Lookup(p.limits.FileTransfer.OwnerPut)
+		if err != nil {
+			return err
+		}
+		uid, err = strconv.Atoi(u.Uid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if p.limits.FileTransfer.GroupPut != "" {
+		u, err := user.LookupGroup(p.limits.FileTransfer.GroupPut)
+		if err != nil {
+			return err
+		}
+		gid, err = strconv.Atoi(u.Gid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if forcedSet {
+		return os.Chown(path, uid, gid)
+	}
+	if p.limits.FileTransfer.PreserveOwner {
+		return os.Chown(path, uid, gid)
+	} else {
+		return nil
+	}
+}
+
+func updatePerHourCounters() {
```

Comment:
The function name is a bit imprecise.

Reply:
crap! and I so renamed all the structs. fixing.